[WIP][FIX] Chat history sync

luckcoolla opened the pull request:
<!-- INSTRUCTION: Keep the line below to notify all core developers about this new PR -->
@RocketChat/android

Closes #1353 #1302 should be checked

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->
I saw that there are some issues with messages synchronization in the chat room during I do turn-off internet and then connects again. Let's check out two described and fixed issues!

**---Issue 1---**
**Steps to Reproduce:**

0. Go to room dialog btw Android client and web-browser client. Turn off internet connection on the android client
1. web-browser client sends the message 
2. Android client written and sends the message
3. Android client connects to the internet
**4. Android client see only own sent messages. web-browser client can see both own and received from Android messages.**

This issue caused by ChatRoomAdapter.kt code (see FIXME message):
<img width="830" alt="screen shot 2018-08-22 at 7 24 50 pm" src="https://user-images.githubusercontent.com/3755157/44592430-3983bc80-a7c9-11e8-857b-d82162b6a661.png">


**---Issue 2---**
**Steps to Reproduce:**

0. Go to room dialog btw Android client and web-browser client. Turn off internet connection on the android client
1. web-browser client sends the message 1
2. Android client written and sends the message 2
3. web-browser client sends the message 3
4. Android client connects to the internet
5. **Open Logcat. Android client asks only for chat history with message 2 and message 3.**

<img width="1150" alt="screen shot 2018-08-22 at 7 44 38 pm" src="https://user-images.githubusercontent.com/3755157/44593234-87012900-a7cb-11e8-8107-64b0bcb0d226.png">


Let's me explain more here:
When user tries to send message without internet, the MessageService : JobService has been started by ChatRoomPresenter#sendMessage method. JobService has own retrySendingMessages loop inside.
Plus there is reconnecting loop in the chat room itself. ChatRoomPresenter#subscribeState waits for reconnected state and calls loadMissingMessages.

Both retry loops(for sending and receiving) are working simultaneously and usually we have next scenario:
When user turns on internet All local unsend messages being sent + saved to local database(so the last saved message for the room is our message). This message received and recorded on the Server as the latest one in the room.
After some time loadMissingMessages() method is working and **brings only our last message + messages after our(if any).**  Because loadMissingMessages call only for newest messages at top of the latest saved message.
So all incoming messages from web-client for this no-internet period in Android client are being ignored.

luckcoolla committed:
Chat history issues fx: on adapter side

luckcoolla committed:
Chat history issues fx: on presenter and repository side

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/Rocket.Chat.Android?pullRequest=1615) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat.Android?pullRequest=1615) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat.Android?pullRequest=1615) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat.Android?pullRequest=1615) it.</sub>

luckcoolla commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -225,6 +226,15 @@ class ChatRoomPresenter @Inject constructor(
+
+        //if success - saving last synced time
+        if (messages.isEmpty()) {
+            messagesRepository.saveLastSyncDate(System.currentTimeMillis())
```
we call saveLastSyncDate() inside loadAndShowMessages() method because it's the plain request for the latest chat history

luckcoolla commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -513,6 +529,9 @@ class ChatRoomPresenter @Inject constructor(
                                 messagesRepository.saveAll(messages.result)
+                                //if success - saving last synced time
+                                // TODO sort and peek the latest date
+                                messagesRepository.saveLastSyncDate(messages.result.last().timestamp)
```
and we call messagesRepository.saveLastSyncDate() inside loadMissingMessages() because here we again are in the point of chat history synchronization

luckcoolla committed:
minor: fixme removed because it fixed

luckcoolla committed:
kotlin NPE on the username assert fixed

luckcoolla committed:
sync date saving updated, checked on device

luckcoolla committed:
loadMissingMessages() improved: unused messages polling from local repository removed

philipbrito commented:
Hi @luckcoolla Could you please sign our CLA?
Thanks for your PR!

philipbrito commented on app/src/main/java/chat/rocket/android/server/infraestructure/MemoryMessagesRepository.kt:
```diff
@@ -7,6 +7,14 @@ import kotlinx.coroutines.experimental.withContext
 
+    private var lastSyncDate: Long? = null
+
+    override suspend fun saveLastSyncDate(it: Long) {
```
It's better to use the same argument name here as defined on the interface: `currentTimeMillis `

luckcoolla replied:
agree

philipbrito commented on app/src/main/java/chat/rocket/android/server/infraestructure/SharedPreferencesMessagesRepository.kt:
```diff
@@ -13,6 +13,28 @@ class SharedPreferencesMessagesRepository(
+        currentServerInteractor.get()?.also { server ->
+            if (!prefs.contains(getSyncDateKey(server)))
+                return@withContext null
+            //
```
Is it supposed to be a comment? 
Could you please remove that or update that?

luckcoolla replied:
yep, removed

philipbrito commented on app/src/main/java/chat/rocket/android/chatroom/adapter/ChatRoomAdapter.kt:
```diff
@@ -150,23 +150,23 @@ class ChatRoomAdapter(
+            return@filter (matchedIndex < 0)
         }
+        //At the second stage we are inserting new received elements into set.
+        if (filteredDataSet.size == 0)
```
Could you please put the `return` in the same line of `if (filteredDataSet.size == 0)`

luckcoolla replied:
sure

luckcoolla committed:
PR #1615 fixes

luckcoolla committed:
Merge branch 'develop' into fix/chat_history_sync

luckcoolla committed:
Merge branch 'develop' into fix/chat_history_sync

luckcoolla commented:
@filipedelimabrito I've covered all points, recheck please)

luckcoolla committed:
Merge branch 'develop' into fix/chat_history_sync

luckcoolla commented:
@filipedelimabrito I've found small issue with unexpected ui behaviour on the attachment file showing.
My changes in ChatRoomAdapter.kt caused it.
I'm gonna fix it in the nearest days, don't merge the PR now please

philipbrito commented:
Thanks for letting us know @luckcoolla . 👍

luckcoolla committed:
ChatRoomAdapter#prependData improved, checked on files, reply, quote, reaction actions

luckcoolla commented:
@filipedelimabrito I've finished changes, now it should be ok in all cases described in this ticket + general messaging cases like quote/reply/attach files.

I caught the main idea that in case of any action like reply/attach/.. UiModelMapper creates several objects: BaseUiModel and then they arrives to ChatRoomAdapter#prependData method.
Just like example if I write new message with reply action, two new message objects are being passed to ChatRoomAdapter#prependData.

In the "#prependData" method I added simple check on new messages time to reduce notifyDataSetChanged() calls 
\+ I added comparison on viewType into the sorting comparator function, so layout on merged chat history should be same as on new messages.

philipbrito committed:
Merge branch 'beta' into fix/chat_history_sync

luciofm commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -184,7 +184,8 @@ class ChatRoomPresenter @Inject constructor(
                         )
                     )
-                    if (oldMessages.isNotEmpty()) {
+                    val lastSyncDate = messagesRepository.getLastSyncDate()
```
This value should be per roomId, otherwise we will use the last sync date from other rooms

luckcoolla replied:
@luciofm yep, it's my mistake. I've tested on my local branch with one room only.
Gonna fix it

luckcoolla replied:
fixed

luckcoolla committed:
PR #1615 fix. Chat synced time associated with roomId

luckcoolla committed:
finally updated ChatRoomAdapter#prependData code: order of merging chat history updated.

luckcoolla commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -483,45 +497,49 @@ class ChatRoomPresenter @Inject constructor(
 
     private fun loadMissingMessages() {
         launch(parent = strategy.jobs) {
-            if (chatRoomId != null && chatRoomType != null) {
```
chatRoomType: String now

luckcoolla committed:
PR #1615 minor change

luckcoolla commented:
@luciofm I've finished code changes.

luciofm committed:
Merge branch 'beta' into fix/chat_history_sync

luciofm commented:
@luckcoolla we are getting a crash when you call prepend on an empty adapter

```
Fatal Exception: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
       at java.util.ArrayList.get(ArrayList.java:437)
       at chat.rocket.android.chatroom.adapter.ChatRoomAdapter.prependData(ChatRoomAdapter.kt:172)
       at chat.rocket.android.chatroom.ui.ChatRoomFragment.showNewMessage(ChatRoomFragment.kt:508)
       at chat.rocket.android.chatroom.presentation.ChatRoomPresenter$loadMissingMessages$1$doResume$$inlined$let$lambda$2.doResume(ChatRoomPresenter.kt:526)
       at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
       at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:149)
       at kotlinx.coroutines.experimental.DispatchedContinuation.run(Dispatched.kt:13)
       at android.os.Handler.handleCallback(Handler.java:789)
       at android.os.Handler.dispatchMessage(Handler.java:98)
       at android.os.Looper.loop(Looper.java:164)
       at android.app.ActivityThread.main(ActivityThread.java:6798)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

luckcoolla commented:
@luciofm yep, I see

`if (minAdditionDate.message.timestamp > this.dataSet[0].message.timestamp) {`

It's the case when you are connecting to the non-empty room from some new device/app installation.

Should I add commit just here in current branch or to create some new pr to beta?

philipbrito commented:
Open a new PR fixing that bug if you can @luckcoolla.
Thank you!

luckcoolla commented:
@filipedelimabrito ok, next time I'll do it in this way)
Now I see @luciofm pushed the crash fix https://github.com/RocketChat/Rocket.Chat.Android/commit/33d3b5f0df0c8fdccb538b3e5d03bdf61a943194
Thanks!
